pkg/server/utils/capacity_filter.go

```diff
@@ -0,0 +1,215 @@
+			workloads = append(workloads, &Reservation{
+				WorkloadName:  cluster.Workers[idx].Name,
+				NeededMemory:  uint64(cluster.Workers[idx].Memory * int(gridtypes.Megabyte)),
+				NeededStorage: uint64(cluster.Workers[idx].DiskSize * int(gridtypes.Gigabyte)),
+			})
+		}
+	}
+
+	err := getNodes(client, workloads)
+	if err != nil {
+		return err
+	}
+
+	if cluster.Master.NodeID == 0 {
+		for _, workload := range workloads {
+			if workload.WorkloadName == cluster.Master.Name {
+				if workload.SelectedNodeID == 0 {
+					return fmt.Errorf("Couldn't find node for workload")
+				}
+				cluster.Master.NodeID = uint32(workload.SelectedNodeID)
+			}
+		}
+	}
+
+	for idx := range cluster.Workers {
+		if cluster.Workers[idx].NodeID == 0 {
+			for _, workload := range workloads {
+				if workload.WorkloadName == cluster.Workers[idx].Name {
+					if workload.SelectedNodeID == 0 {
+						return fmt.Errorf("Couldn't find node for workload")
+					}
+					cluster.Workers[idx].NodeID = uint32(workload.SelectedNodeID)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// Assign chosen NodeIds to machines vm. with both way conversions to/from Reservations array.
+func AssignNodesIDsForMachines(client *deployer.TFPluginClient, machines *types.MachinesModel) error {
+	// all units unified in bytes
+
+	workloads := []*Reservation{}
+
+	for idx := range machines.Machines {
+		if machines.Machines[idx].NodeID == 0 {
+			neededStorage := 0
+			for _, disk := range machines.Machines[idx].Disks {
+				neededStorage += disk.SizeGB * int(gridtypes.Gigabyte)
+			}
+			for _, qsfs := range machines.Machines[idx].QSFSs {
+				neededStorage += int(qsfs.MaxZDBDataDirSize) * int(gridtypes.Gigabyte)
+			}
+			neededStorage += machines.Machines[idx].RootfsSize * int(gridtypes.Megabyte)
+
+			workloads = append(workloads, &Reservation{
+				WorkloadName:  machines.Machines[idx].Name,
+				NeededMemory:  uint64(machines.Machines[idx].Memory * int(gridtypes.Megabyte)),
+				NeededStorage: uint64(neededStorage),
+			})
+		}
+	}
+
+	err := getNodes(client, workloads)
+	if err != nil {
+		return err
+	}
+
+	for idx := range machines.Machines {
+		if machines.Machines[idx].NodeID == 0 {
+			for _, workload := range workloads {
+				if workload.WorkloadName == machines.Machines[idx].Name {
+					if workload.SelectedNodeID == 0 {
+						return fmt.Errorf("Couldn't find node for workload")
+					}
+					machines.Machines[idx].NodeID = uint32(workload.SelectedNodeID)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func getFreeResources(node proxyTypes.Node) proxyTypes.Capacity {
+	return proxyTypes.Capacity{
+		MRU: (node.TotalResources.MRU - node.UsedResources.MRU),
+		SRU: (node.TotalResources.SRU - node.UsedResources.SRU),
+	}
+}
+
+func checkNodeAvailability(node proxyTypes.Node, workload Reservation, reservedCapacity map[uint64]Reservation) bool {
+	if uint64(getFreeResources(node).MRU) >= reservedCapacity[uint64(node.NodeID)].NeededMemory+uint64(workload.NeededMemory) &&
```

Comment:
assign node free resources first, then check on the result:
free := getFreeResources(node)
free.mru, free.sru